### ISSUE TITLE
Using the `/bbb` command doesn't create meeting

### ISSUE THREAD

**Issue by jdanders** [blindsidenetworks/mattermost-plugin-bigbluebutton#32]
Maybe related to issue #27.

I am running Mattermost 5.6.2 (bundled with gitlab) and recently did a clean install of BBB. I am able to create and join meeting by hand using the api. When I do `/bbb ` in a chat channel and then click on "Join Meeting" I get:
`Error: The meeting ID that you supplied did not match any existing meetings` on the default splash screen of BBB.

In watching the `nginx` logs on the BBB server, it's pretty clear that the plugin is not trying to create the meeting. I grabbed the `meetingID` and `password` out of the failed `join` call and created a meeting with the attempted `meetingID` using a direct api call. After I did the `create` api call, the "Join Meeting" button in Mattermost worked just fine.

Am I missing something in my BBB configuration that allows a `join` to implicitly create a meeting? If not, then it seems like the plugin is failing to do a `create` call before trying to join.

**Comment by ppi1302** [comment]
Hello,
any news on that as I have the same problem with the latest 2.0.2 Mattermost  plugin and BBB 2.0 RC11
Many thanks in advance

**Comment by harshilsharma63** [comment]
Hi @jdanders , @ppi1302 do you still face this issue in a more recent version of Mattermost? I just tried on Mattermost v5.13.2 and it worked fine.

Could you please try upgrading your Mattermost instances?

**Comment by harshilsharma63** [comment]
@jdanders @ppi1302 I'm closing this issue as the bug is not reproducible as of now. Feel free to re-open this open open a new issue if you face the problem again.

Thanks!

**Comment by wyrdnixx** [comment]
Same problem here.
mattermost with docker-compose:
Mattermost Version: 5.14.1
Database Schema Version: 5.14.0

BigBlueButton 
dedicated VM-Server 
Version 1:2.2.0-4

Local LAN Setup / no firewall

ngnix log of bbb-server only shows a "join" when i try to connect.
192.168.X.XX - - [21/Sep/2019:21:06:55 +0200] "GET /bigbluebutton/api/join?fullName=XXXXXX&meetingID=1R9wqd5WfkG7M25abdbuRJnD2AE&password=mp&joinViaHtml5=true&checksum=17c88adf881340d53XXXXXXXXXX2018c8fd0d5 HTTP/1.1" 302 0 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:69.0) Gecko/20100101 Firefox/69.0"

when i create the bbb there is no call / nothing on the nginx log

**Comment by wyrdnixx** [comment]
> 
> 
> Same problem here.
> mattermost with docker-compose:
> Mattermost Version: 5.14.1
> Database Schema Version: 5.14.0
> 
> BigBlueButton
> dedicated VM-Server
> Version 1:2.2.0-4
> 
> Local LAN Setup / no firewall
> 
> ngnix log of bbb-server only shows a "join" when i try to connect.
> 192.168.X.XX - - [21/Sep/2019:21:06:55 +0200] "GET /bigbluebutton/api/join?fullName=XXXXXX&meetingID=1R9wqd5WfkG7M25abdbuRJnD2AE&password=mp&joinViaHtml5=true&checksum=17c88adf881340d53XXXXXXXXXX2018c8fd0d5 HTTP/1.1" 302 0 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:69.0) Gecko/20100101 Firefox/69.0"
> 
> when i create the bbb there is no call / nothing on the nginx log



-> i think i found the issue.
i run the test-install of bbb with an selfsigned ssl certificate

the log from mattermost shows 

{"level":"info","ts":1569093367.6986592,"caller":"go-plugin/stream.go:15","msg":"2019/09/21 21:16:07 HTTP GET ERROR: Get https://bigbluebutton.XXXXXX/bigbluebutton/api/isMeetingRunning?meetingID=1R9v5xKwR7tciM0L4yoDvfuSO08&checksum=4690913f42945203cb2544945c0b92015098dc31: x509: certificate signed by unknown authority","plugin_id":"bigbluebutton","source":"plugin_stderr"}

Certificate Issue

**Comment by jdanders** [comment]
Sounds like this issue should be re-opened.

**Comment by ppi1302** [comment]
Should be, Still same issue with Mattermost 5.15

**Comment by ppi1302** [comment]
Please reopen the case as it is not yet solved. Thanks

**Comment by harshilsharma63** [comment]
@jdanders @ppi1302 thanks for reporting the error. I'm re-opening the issue and we'll investigate on this.

**Comment by Fox3150** [comment]
@jdanders @ppi1302 This potentially could resolve your issue. Check your mattermost.log in /opt/mattermost/logs/

I am running an on-site setup of both mattermost and bigbluebutton and also am using an internal Certificate Authority for SSL. I was experiencing the exact same issue with it not creating a room, after some investigating and checking log files I found this:
`{"level":"info","ts":1572720791.0277526,"caller":"go-plugin/stream.go:15","msg":"2019/11/02 18:53:11 HTTP GET ERROR: Get https://SERVER/bigbluebutton/api/getRecordings?meetingID=1T4XDZE2cXbhzJHwPVKmfCBcRMx&checksum=f51e08da38f3228e3d1ad882efc790565ed11193: x509: certificate signed by unknown authority","plugin_id":"bigbluebutton","source":"plugin_stderr"}`

I added the Internal RootCA to my Mattermost server and performed a reboot and presto:
`{"level":"info","ts":1572720947.7459471,"caller":"go-plugin/stream.go:15","msg":"2019/11/02 18:55:47 SUCCESS CREATE MEETINGROOM. MEETING ID: 1T4YfxUVfGH8klA3mT3BszbaKil","plugin_id":"bigbluebutton","source":"plugin_stderr"}`

If you are using https it's possible that this might be your issue. I'm fairly new to these systems as I only started their implementation in the last two weeks. @ffdixon is https required for an API call to create the meeting? If it is not you could make sure it's just http in your BigBlueButton plugin settings on Mattermost. Hope this helps.

**Comment by ffdixon** [comment]
We would recommend using HTTPS for the API calls.  

> x509: certificate signed by unknown authority

Is this BigBlueButton server running a self-signed certificate?  You need a proper certificate issued by a CA authority (such as Let's Encrypt).

**Comment by Fox3150** [comment]
@ffdixon As these are internal only we are running a Windows Enterprise CA (Due to the fact we don't use public domain suffixes) for certificates on all internal systems. We push the Trusted Root to our Windows machines via Group Policy however that obviously does not work for Linux. I simply needed to add the Root Certificate to be implicitly trusted on both the Mattermost and the BigBlueButton servers.

Since HTTPS is recommended for the API calls, I would recommend that @jdanders @ppi1302  check the mattermost logs as they might have information on why the API call is not succeeding.

**Comment by jdanders** [comment]
@ffdixon -- this is an entirely self-contained network and we have self-signed certs. As a self-hosted solution, dependency on an external authority like Let's Encrypt shouldn't be a requirement.

@Fox3150 -- I'm not sure what I'm seeing. I do see this in my mattermost logs:
```
2019-11-19_19:28:11.65133 {"level":"info","ts":1574191691.6511877,"caller":"go-plugin/stream.go:15","msg":"2019/11/19 12:28:11 HTTP GET ERROR: Get https://SERVER/bigbluebutton/api/isMeetingRunning?meetingID=1GuNoDFBzIf2GItnCTjgsqOh&checksum=fa8e49efb0d1ecfabb470d95623d1e44372: x509: certificate signed by unknown authority","plugin_id":"bigbluebutton","source":"plugin_stderr"}
2019-11-19_19:28:11.65136 {"level":"info","ts":1574191691.651288,"caller":"go-plugin/stream.go:15","msg":"2019/11/19 12:28:11 ERROR: HTTP ERROR.","plugin_id":"bigbluebutton","source":"plugin_stderr"}
```
But the cert is installed at the OS level. If I do `curl --verbose https://SERVER/` I get:
```
* Connected to SERVER (192.1.1.13) port 443 (#0)
* found 148 certificates in /etc/ssl/certs/ca-certificates.crt
* found 593 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
* 	 server certificate verification OK
```
So it seems like the OS running gitlab and mattermost (omnibus install) does in fact trust the root CA, but the plugin's HTTP GET doesn't.

Also, my logs have this, and I'm not sure how important this is:
```
2019-11-19_19:28:11.62832 {"level":"debug","ts":1574191691.6282055,"caller":"mlog/log.go:161","msg":"CSRF Check failed for request - Please migrate your plugin to either send a CSRF Header or Form Field, XMLHttpRequest is deprecated","path":"/plugins/bigbluebutton/ismeetingrunning","ip":"127.0.0.1:48460","session_id":"rygbjkxjitff4r674xn9s4r","user_id":"67u9iya4o3fezf5u1rgk44e"}
```

**Comment by Fox3150** [comment]
@jdanders When you say you have self-signed certificates, are those certificates by a single root authority or is each machine signing its own?

In the instance that each machine is signing its own, you will need to take the certificate from each of the machines and transfer onto the other. Depending on your linux distribution it could be in different locations but for debian: /usr/local/share/ca-certificates then run update-ca-certificates.

You may only need to put the BBB cert onto the Mattermost machine, I did both just to be certain. Then obviously you will need both certificates on all of your client machines as well.

**Comment by jdanders** [comment]
@Fox3150 I mean a single root authority. 

The `curl` test shows that the Mattermost machine (as client) can talk to the BBB machine (as server) using the certificates.

**Comment by jdanders** [comment]
Okay, so I'm back to trying to get this to work again. I am using Gitlab Mattermost with the omnibus install, and it turns out they have a separate copy of trusted certs, so I put my self-signed cert in `/etc/gitlab/trusted-certs/` and after restarting gitlab, it seems to be accepting the cert now. 

However, it's still not working;
```
2020-04-09_04:56:33.75335 {"level":"debug","ts":1586408193.7532322,"caller":"mlog/log.go:162","msg":"CSRF Check failed for request - Please migrate your plugin to either send a CSRF Header or Form Field, XMLHttpRequest is deprecated","path":"/plugins/bigbluebutton/ismeetingrunning","ip":"127.0.0.1:35516","session_id":"jiac9pyh6pgxbj83g1zej39g7e","user_id":"67u9iya4o3fp8fezf5u1rgk44e"}
2020-04-09_04:56:34.03694 {"level":"debug","ts":1586408194.036799,"caller":"web/handlers.go:77","msg":"Received HTTP request","method":"POST","url":"/api/v4/users/status/ids","request_id":"dz9d6pnoofrpfrpb9iuxbtiykw"}
2020-04-09_04:56:34.44747 {"level":"debug","ts":1586408194.4473016,"caller":"mlog/log.go:162","msg":"CSRF Check failed for request - Please migrate your plugin to either send a CSRF Header or Form Field, XMLHttpRequest is deprecated","path":"/plugins/bigbluebutton/joinmeeting","ip":"127.0.0.1:35520","session_id":"jiac9pyh6pgxbj83g1zej39g7e","user_id":"67u9iya4o3fp8fezf5u1rgk44e"}
2020-04-09_04:56:34.45542 {"level":"error","ts":1586408194.4553683,"caller":"mlog/sugar.go:23","msg":"CREATE MEETINGROOM FAILD: <Map><timestamp>1586408194456</timestamp><status>404</status><error>Not Found</error><message>No message available</message><path>/bigbluebutton/create</path></Map>","plugin_id":"bigbluebutton"}
```

Any tips on how to debug this?

**Comment by rottaran** [comment]
My guess because ip:127.0.01 looks strange: The server side of the plugin has to connect to bbb in order to set up the meeting etc. This connection is encrypted or should be encrypted. The request has to be addressed to the public domain name of the bbb server, not just localhost. Otherwise the bbb side cannot accept the connection. Did you use the full servername in your plugin configuration?

**Comment by jdanders** [comment]
I noticed that too, but yes I used the full servername in the config, and I do see traffic going from gitlab/mattermost server to BBB server. However, when I checked the servername I noticed that the instructions below the URL entry box say "`Enter the url of your BigBlueButton server ending with /bigbluebutton/api`". I only had it up to `/bigbluebutton` so I was missing the `api` component. Adding `api` fixed it! Now it's like this:
`https://internal.ip.net/bigbluebutton/api`

These logs are indeed from the client side Mattermost log. In the omnibus install, it's located here for others trying to debug: `/var/log/gitlab/mattermost/current`
